api/src/pcapi/routes/auth/discord.py

```diff
@@ -1,14 +1,100 @@
+import datetime
+from urllib import parse as urlparse
+from urllib.parse import urlencode
+
+from flask import redirect
 from flask import render_template
 from flask import request
+from werkzeug.wrappers.response import Response
 
+from pcapi import settings
+from pcapi.app import csrf
+from pcapi.core import token as token_utils
+from pcapi.core.users import exceptions as users_exceptions
+from pcapi.core.users import models as user_models
+from pcapi.core.users import repository as users_repo
+from pcapi.models import db
 from pcapi.routes.auth.forms.forms import SigninForm
 
 from . import blueprint
 
 
+def process_redirect_url(redirect_url: str, authorization_code: str) -> str:
+    # Prepare the redirect URL
+    url_parts = list(urlparse.urlparse(redirect_url))
+    queries = dict(urlparse.parse_qsl(url_parts[4]))
+    queries.update({"authorization_code": authorization_code})
+    url_parts[4] = urlencode(queries)
+    url = urlparse.urlunparse(url_parts)
+    return url
+
+
 @blueprint.auth_blueprint.route("/discord/signin", methods=["GET"])
 def discord_signin() -> str:
     form = SigninForm()
     form.redirect_url.data = request.args.get("redirect_url")
     form.discord_id.data = request.args.get("discord_id")
     return render_template("discord_signin.html", form=form)
+
+
+@blueprint.auth_blueprint.route("/discord/signin", methods=["POST"])
+def discord_signin_post() -> str | Response | None:
+    csrf.protect()
+    form = SigninForm()
+    if not form.validate():
+        form.error_message = "Identifiant ou Mot de passe incorrect"
+        return render_template("discord_signin.html", form=form)
+    email = form.email.data
+    password = form.password.data
+    url_redirection = form.redirect_url.data
+    discord_id = form.discord_id.data
+    try:
+        user = users_repo.get_user_with_credentials(email, password, allow_inactive=True)
+    except users_exceptions.UnvalidatedAccount:
+        form.error_message = "L'email n'a pas été validé"
+        return render_template("discord_signin.html", form=form)
+
+    except users_exceptions.CredentialsException:
+        form.error_message = "Identifiant ou Mot de passe incorrect"
+        return render_template("discord_signin.html", form=form)
+
+    if user.account_state.is_deleted:
+        form.error_message = "Le compte a été supprimé"
+        return render_template("discord_signin.html", form=form)
+
+    if user.account_state == user_models.AccountState.ANONYMIZED:
+        form.error_message = "Le compte a été anonymisé"
+        return render_template("discord_signin.html", form=form)
+    discord_user = user.discordUser
+    if discord_user is None or discord_user.has_access is False:
```

Comment:
[détail / style] Plus pythonique : 

```suggestion
    if discord_user is None or not discord_user.has_access:
```